src/ratelimit.ts

```diff
@@ -148,37 +165,64 @@ export abstract class Ratelimit<TContext extends Context> {
    */
   public limit = async (
     identifier: string,
-    req?: { geo?: Geo; rate?: number },
+    req?: {
+      geo?: Geo,
+      rate?: number,
+      ip?: string,
+      userAgent?: string,
+      country?: string
+    },
   ): Promise<RatelimitResponse> => {
     const key = [this.prefix, identifier].join(":");
     let timeoutId: any = null;
+
+    const members = [identifier, req?.ip, req?.userAgent, req?.country];
+    const definedMembers = members.filter(item => item !== undefined) as string[];
     try {
-      const arr: Promise<RatelimitResponse>[] = [this.limiter().limit(this.ctx, key, req?.rate)];
-      if (this.timeout > 0) {
-        arr.push(
-          new Promise((resolve) => {
-            timeoutId = setTimeout(() => {
-              resolve({
-                success: true,
-                limit: 0,
-                remaining: 0,
-                reset: 0,
-                pending: Promise.resolve(),
-              });
-            }, this.timeout);
-          }),
-        );
+      let res: RatelimitResponse;
+      const deniedMember = checkDenyListCache(definedMembers)
+      if (deniedMember) {
+        res = defaultDeniedResponse(deniedMember);
+      } else {
+        const arr: Promise<[RatelimitResponse, string | undefined]>[] = [Promise.all([
+          this.limiter().limit(this.ctx, key, req?.rate),
+          checkDenyList(
+            this.primaryRedis,
+            this.prefix,
+            definedMembers
+          )
+        ])];
+        if (this.timeout > 0) {
+          arr.push(
+            new Promise((resolve) => {
+              timeoutId = setTimeout(() => {
+                resolve([
+                  {
+                    success: true,
+                    limit: 0,
+                    remaining: 0,
+                    reset: 0,
+                    pending: Promise.resolve(),
+                    reason: "timeout"
+                  },
+                  undefined
+                ]);
+              }, this.timeout);
+            }),
+          );
+        }
+        res = resolveResponses(...(await Promise.race(arr)));
```

Comment:
I think its time to break this `limit()` function into smaller pieces. Its starting to get really hard to follow. Also, having some explanatory comments would be fantastic, wdyt? We don't have do this right away.

Reply:
I also thought that it was getting out of control. Did some refactoring and I think now it looks way cleaner.

Reply:
10x better. I like it.